src/core/router/index.php

```diff
@@ -0,0 +1,142 @@
+        $this->sortRoutes();
+    }
+
+    public function put($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'PUT', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function patch($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'PATCH', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function delete($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'DELETE', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    private function findSegments($route, $segmentsRoute, $segmentsUrl) {
+        if (count($segmentsRoute) !== count($segmentsUrl)) return null;
+
+        foreach ($segmentsRoute as $i => $segment) {
+            if (strpos($segment, ':') === 0) {
+                $key = substr($segment, 1);
+                $route['params'][$key] = $segmentsUrl[$i];
+            } else {
+                if ($segment !== $segmentsUrl[$i]) return null;
+            }
+        }
+
+        return $route;
+    }
+
+    private function findRoute($req) {
+        $pathname = strtok($req['url'], '?');
+        $segmentsUrl = explode('/', $pathname);
+
+        foreach ($this->routes as $route) {
+            $newRoute = $route;
+            $newRoute['params'] = [];
+            $segmentsRoute = explode('/', $route['route']);
+            $segments = $this->findSegments($newRoute, $segmentsRoute, $segmentsUrl);
+            if ($segments) return $newRoute;
+        }
+
+        return null;
+    }
+
+    private function getAgent($req) {
+        return $req['headers']['User-Agent'] ?? null;
+    }
+
+    private function getCookies($req) {
+        if (!isset($req['headers']['Cookie'])) return null;
+
+        $cookies = [];
+        foreach (explode('; ', $req['headers']['Cookie']) as $pair) {
+            list($key, $value) = explode('=', $pair, 2);
+            $cookies[$key] = $value;
+        }
+
+        return $cookies;
+    }
+
+    private function getParams($req) {
+        $query = [];
+        parse_str(parse_url($req['url'], PHP_URL_QUERY), $query);
+        return $query;
+    }
+
+    private function getBody($req) {
+        /* TODO:const body = await formidable(req);
+        return body; */
+    }
+
+    private function getIp($req) {
+        return $req['headers']['X-Forwarded-For'] ?? $req['remote_addr'];
+    }
+
+    public function request($req, $res) {
+        $pathname = strtok($req['url'], '?');
+        $route = $this->findRoute($req);
+        $cookies = $this->getCookies($req);
+        $params = $this->getParams($req);
+        $body = $req['method'] === 'POST' ? $this->getBody($req) : [];
+        $agent = $this->getAgent($req);
+        $ip = $this->getIp($req);
+
+        $ctx = ['params' => array_merge(['_state' => ['pathname' => $pathname, 'cookie' => $cookies, 'route' => $route, 'agent' => $agent, 'ip' => $ip]], $params, $body)];
```

Comment:
It's not critical, but try to store individual arrays in separate variables to avoid long lines of code. If the array is large, use line breaks.

You can learn more here:
https://www.php-fig.org/psr/psr-12/

<img width="755" alt="Screenshot 2024-12-28 at 09 36 52" src="https://github.com/user-attachments/assets/6f6fc09f-603d-4cb5-b767-b29dd7656b3f" />

Reply:
Done